odb_extractor_atoms.py

```diff
@@ -59,98 +223,297 @@ def select_obsclass(classes):
 # --------------
 
 
-def findatoms(observation):
-    '''Analyze a json observing sequence from the ODB and define atoms.'''
+def autocorr_lag(x, plot=False):
+    '''Test for patterns with auto-correlation'''
+
+    # Auto correlation
+    result = np.correlate(x, x, mode='full')
+    corrmax = np.max(result)
+    if corrmax != 0.0:
+        result = result / corrmax
+    #     print(result)
+    #     print(result.size, result.size//2)
+    if plot:
+        plt.plot(result[result.size // 2:])
+
+    # Pattern offset using first prominent peak
+    # lag = np.argmax(result[result.size//2 + 1:]) + 1
+    peaks, prop = find_peaks(result[result.size // 2:], height=(0, None), prominence=(0.25, None))
+    #     print(peaks)
+    #     print(prop)
+    lag = 0
+    if len(peaks) > 0:
+        lag = peaks[0]
+
+    return lag
+
+# --------------
+
+
+def findatoms(observation, verbose=False, ws=None, fid=sys.stdout):
+    """Analyze a json observing sequence from the ODB and define atoms."""
 
     classes = []
+    guiding = []
     qastates = []
     atoms = []
     natom = 0
-    nabba = 0
 
     # Make dictionary out of obslog to get QA state
     obslog = {}
-    for log_entry in observation['obsLog']:
-        obslog[log_entry['label']] = {'qaState': log_entry['qaState'], 'filename': log_entry['filename']}
-    datalabels = list(obslog.keys())
+    datalabels = []
+    if 'obsLog' in observation.keys():
+        for log_entry in observation['obsLog']:
+            obslog[log_entry['label']] = {'qaState': log_entry['qaState'], 'filename': log_entry['filename']}
+        datalabels = list(obslog.keys())
 
     # Sequence analysis
     sequence = observation['sequence']
     nsteps = len(sequence)
 
+    # First pass to analyze offsets and exptimes/coadds
+    do_not_split = False
+    exptimes = []
+    coadds = []
+    qoffsets = []
+    poffsets = []
+    qoffsets_sky = []
+    poffsets_sky = []
+    config = {'inst': '', 'fpu': [], 'disperser': [], 'filter': [], 'wavelength': []}
+    for ii, step in enumerate(sequence):
+        step_keys = list(step.keys())
```

Comment:
Just a Python coding tip: we should not have to do this, and taking a `dict_keys` object and converting it to a  `list` is (probably) quite inefficient: I am not completely certain of how Python handles this, but I imagine that it is an `O(n)` operation both in execution and memory use where `n` is the number of entries in `step`.

We can just ask:
```
if 'instrument:disperser' in step:
     ...
```
which automatically looks at the keys only to see if they contain `'instrument_disperser'` instead of doing:
```
step_keys = list(step.keys())
...
if 'instrument:disperser' in step_keys:
    ....
```

Reply:
Thanks, I'll try to remember that